Move to java.util.concurrent.locks.Lock (Was: Add concurrent.Lock#doWith)

lcycon opened the pull request:
A method I have used on a few occasions to eliminate the boilerplate acquire/release when executing simple statements in a locked environment.

Review by @phaller please!

Commit:
Add Lock#doWith to wrap acquire/release logic

retronym commented:
I would actually vote to deprecate `scala.concurrent.Lock`. Why on earth is `available` public?!?

lcycon commented:
I was wondering what was going on with `available`.  Seemed to me an odd idea to have that piece of mutable state exposed.

I don't see deprecating `Lock` myself. I can imagine situations where one would want a simple lock in lieu of other more heavyweight concurrency structures.

phaller commented:
The `Lock` interface of the `java.util.concurrent.locks` package is a bit more flexible than `scala.concurrent.Lock` with methods like `tryLock` and `newCondition`. Maybe it would be better to add something like `doWith` as an extension method to the Java `Lock`, and deprecate `scala.concurrent.Lock`.

lcycon commented:
Hm.  That does sound promising.  I'll hack around on that idea after work.

Commit:
Move to java.util.concurrent.locks.Lock

- Add deprecation notice to scala.concurrent.Lock
- Add RichLock type to support extension methods
  - Implicit conversion to RichLock added to Predef

lcycon commented:
@phaller Here is a change that encapsulates that idea.  I like this route much better than tacking on to the under-featured `scala.concurrent.Lock`, seems more useful ^.^

soc commented:
I'm sceptical that a) `scala.runtime` is the right place for the decorator and b) adding yet-another-implicit to `Predef` is the way to go.

Opinions?

lcycon commented:
@soc I was most definitely unsure about where to put the decorator.  I followed suit with some of the other runtime class decorators.  Feel free to point me in the right direction!

Is there somewhere else that would give us an import-free implicit conversion to the decorator?  Or are you arguing for some import based conversion?

retronym commented:
I don't think this one is worth adding. I sort of doubt that it will be discovered by enough potential users to warrant inclusion. If it were part of a larger set of decorators for `java.util.concurrent`, this would be less of an issue. But that sort of library would be better developed as an independent library initially.

gkossakowski commented:
I agree with @retronym.

retronym commented:
I would accept a PR to deprecate that dodgy `Lock` class, though.

lcycon commented:
Cool, fair enough.  Pretty small decorator anyway ;)

See #2780 for the deprecation stuff
